### PR TITLE
Wip/cbl/11 10

### PR DESCRIPTION
**Pull request by devbyaccident**

### CODE REVIEW

File: exercises/00-Environment-Setup/README.md [hunk 1]
```diff
@@ -33,41 +33,15 @@ docker --version
...
-*   Defines a service called `jenkins`.
-*   Uses the official Jenkins server running JDK17 (`lts-jdk17`) images from Docker Hub.
-*   Allows the container to run with root access to devices on the host machine. 
-**    You will need this for Jenkins to be able to build containers later on, but it's usually a good idea to avoid running contianers in privelaged mode.
-*   Restarts the container if it fails.
-*   Maps port 8080 on the host machine to port 8080 in the container.
-*   Volumes the `/var/jenkins_home` directory within the container for persistent data and maps the host's `/var/run/docker.sock` file as a volume to enable communication between containers.
-*   Makes a volume called `jenkins_home` available inside the container.
+- Defines services for Jenkins, SonarQube, and a Jenkins agent. Don't worry about what these are yet, you'll learn more abou them as you go through the course.
+-Creates a shared network so these services can connect to each other.
```
Comment (chris-s-friedman) [inline exercises/00-Environment-Setup/README.md]:
```suggestion
- Creates a shared network so these services can connect to each other. 
```

need the space there so it renders correctly

File: exercises/00-Environment-Setup/README.md [hunk 2]
```diff
@@ -93,12 +67,14 @@ Navigate to `http://localhost:8080` in your web browser. Follow these steps:
 1.  **Enter the initial admin password**: You can get this from the jenkins container by running the command below.
 
 ```bash
-sudo docker exec jenkins-jdk-17 cat /var/jenkins_home/secrets/initialAdminPassword
+docker exec jenkins cat /var/jenkins_home/secrets/initialAdminPassword
 ```
 
 2.  **Install the Reccomended Plugins**: You'll install more later, but these are a good start for now.
```
Comment (chris-s-friedman) [inline exercises/00-Environment-Setup/README.md:73]: Installing these got a failure first time around 
image

Retrying 3 times finally got everything installed

File: exercises/00-Environment-Setup/README.md [hunk 3]
```diff
@@ -93,12 +67,14 @@ Navigate to `http://localhost:8080` in your web browser. Follow these steps:
...
 3.  **Create a new Admin user**: Enter your details, choose an admin password, and save.
 
 Congratulations! You have now set up a Jenkins server running on Docker Compose. This is a basic example to get you started with the exercises locally, and you will be learning more about this setup and making changes to it throughout the course.
 
+To shut these down at anytime, run `docker compose down` from this directory. 
+
 For now, go ahead and let the instructor know you're done.
```
Comment (chris-s-friedman) [inline exercises/00-Environment-Setup/README.md:82]: after doing the admin account setup, i had this page. there's no instructions on what to do from this step 

image

File: exercises/01-Installing-Plugins/README.md
Comment (chris-s-friedman) [inline exercises/01-Installing-Plugins/README.md:1]: after navigating to the manage jenkins page, I had a list of plugins to install 
image

**Review by chris-s-friedman — COMMENTED** [review]
Still working on review, got to the manage jenkins page in step 01 and had some errors, stopped there for the evening

File: exercises/01-Installing-Plugins/README.md
Comment (chris-s-friedman) [reply exercises/01-Installing-Plugins/README.md:1]: just a note on this one: ignoring this didn't cause a problem (so far - at least in completing the instructions on step 1)

File: exercises/01-Installing-Plugins/README.md [hunk 4]
```diff
@@ -46,4 +48,13 @@ Navigate to http://localhost:8080/restart and click **Yes**, or check the box la
...
 Once Jenkins restarts, you'll be back to the sign in screen and you'll be able to use the new plugin.
 
+**Step 4: Install Other Course-Required Plugins**
+-------------------------------------------
+
+Now that you know how to install plugins, here's some of the other ones we're going to use during the course. Make sure these are installed before finishing the exercise.
+
+**SonarQube Scanner for Jenkins** 
+**GitHub Integration**
+**GitHub Branch Source**
```
Comment (chris-s-friedman) [inline exercises/01-Installing-Plugins/README.md]:
```suggestion
- **SonarQube Scanner for Jenkins** 
- **GitHub Integration**
- **GitHub Branch Source**
```

Don't know if you want a bulleted list, but these rendered all on one line, so recommending a list so they are all on seperate lines

Comment (chris-s-friedman) [reply exercises/01-Installing-Plugins/README.md]: Also, maybe you can link to their page in the jenkins plugin directory, so we know we have the correct one

File: exercises/01-Installing-Plugins/README.md
Comment (chris-s-friedman) [reply exercises/01-Installing-Plugins/README.md:1]: and another note: after restarting jenkins, all of these failures disapeared

File: exercises/01-Installing-Plugins/README.md [hunk 5]
```diff
@@ -46,4 +48,13 @@ Navigate to http://localhost:8080/restart and click **Yes**, or check the box la
 
 Once Jenkins restarts, you'll be back to the sign in screen and you'll be able to use the new plugin.
 
+**Step 4: Install Other Course-Required Plugins**
+-------------------------------------------
+
+Now that you know how to install plugins, here's some of the other ones we're going to use during the course. Make sure these are installed before finishing the exercise.
+
+**SonarQube Scanner for Jenkins**
```
Comment (chris-s-friedman) [inline exercises/01-Installing-Plugins/README.md]: searching `SonarQube Scanner for Jenkins` doesn't eturn any results, but ommitting `for jenkins` shows your plugin.

image

File: exercises/01-Installing-Plugins/README.md (hunk 4 above)
Comment (chris-s-friedman) [inline exercises/01-Installing-Plugins/README.md]: looks like `GitHub Branch Source` is not the ful name of your plugin

image

File: exercises/02-Jenkins-Projects/README.md
Comment (chris-s-friedman) [inline exercises/02-Jenkins-Projects/README.md:1]: In step 1: 

> Navigate to the Jenkins console

What is the `console`? Is that the dashboard?

Comment (chris-s-friedman) [reply exercises/02-Jenkins-Projects/README.md:1]: in step 3, spelling error on `Aggrigating`

File: exercises/02-Jenkins-Projects/README.md [hunk 6]
```diff
@@ -70,10 +71,47 @@ To go a step further, click on **Console Output** on the left side of the screen
 
 > You remember when I said the shell commands had a problem? Can you see what that problem is and why it happened?
```
Comment (chris-s-friedman) [inline exercises/02-Jenkins-Projects/README.md:72]: is it that we haven't configured git yet?

File: exercises/03-Polling-SCM/README.md [hunk 7]
```diff
@@ -1,8 +1,96 @@
...
 
+Like most CI build systems, Jenkins is intended to be used to trigger builds when new changes are found in Source Code Management (SCM).
 
-**Step 1: Setup a Freestyle Project**
+In this exercise, you'll setup a Git repository and poll it for changes. If any are made, it will trigger the jobs you made in the last exercise.
+
+**Step 1: Setup a Git Repository**
+-------------------------------------------
+
+First, you'll need to have a Git repository to make changes to. You're free to make one of your own, or clone the [Demo Repo from my account](https://github.com/devbyaccident/Demo_Repo).
```
Comment (chris-s-friedman) [inline exercises/03-Polling-SCM/README.md]: What caught my eye was using the word `clone` - are you using that for an audience that would be unfamiliar with the term `fork`?

The other thing I'm thinking about - and take this with the grain of salt that I'm commenting before reading the rest of this page - shouldn't this really say `git repo on github` instead of just `git repo`? just trying to draw the distinction/ make sure it's clear that you're not talking about a locally cloned repo (yet) and instead you're talking about something on github

File: exercises/03-Polling-SCM/README.md [hunk 8]
```diff
@@ -1,8 +1,96 @@
...
+> Note: If you aren't comfortable using Git, please let the instructor know so they can walk through the steps required.
+
+**Step 2: Setup a Github Access Token**
 -------------------------------------------
 
+Unless you only want to poll public repositories, and only under Github's rate limit, you're going to need to authenticate your git calls. There are many ways to do this, but for demo purposes, you'll be setting up an simple one today that doesn't require installing an application in Github.
+
+While logged into your Github account, go to https://github.com/settings/tokens.
+
+From here, navigate to **Personal access tokens (classic)** and create a new token. The token should have the scopes below:
```
Comment (chris-s-friedman) [inline exercises/03-Polling-SCM/README.md:24]: There's two types of tokens to generate for personal access tokens (classic) 
image

File: exercises/03-Polling-SCM/README.md [hunk 9]
```diff
@@ -1,8 +1,96 @@
...
+- *admin:repo_hook* - for managing hooks (read, write and delete old ones)
+- *repo* - to see private repos
+- *repo:status* - to manipulate commit statuses
+
+Once the token is created, copy it to notepad for use in the next step.
+
+**Step 3: Setup a Jenkins Credential**
+-------------------------------------------
+
+Jenkins has it's own credential store for secrets that need to be used in build pipelines. To add it to Jenkins, go to the Jenkins home page, then click on **Manage Jenkins** > **Security** > **Credentials**
```
Comment (chris-s-friedman) [inline exercises/03-Polling-SCM/README.md]: > jenkins home page

is this the dashboard?

File: exercises/03-Polling-SCM/README.md [hunk 10]
```diff
@@ -1,8 +1,96 @@
...
+- Username: Your Github username
+- Password: The Github access token you created in step 2
+- ID: `Github-credentials`
+
+Once that's done, click **Create**
+
+**Step 4: Poll Github from a Freestyle Job**
+-------------------------------------------
+
+Now that the credentials are setup, it's time to use them. Open up your Hello World Freestyle job and open the configure page.
```
Comment (chris-s-friedman) [inline exercises/03-Polling-SCM/README.md]: Just a note: when we created the job, we named it `Hello_World_Project`

File: exercises/03-Polling-SCM/README.md [hunk 11]
```diff
@@ -1,8 +1,96 @@
...
+
+Under the **Source Code Management** section, choose **Git**, then fill out the rest of the fields with the values below:
+
+- Repository URL: The HTTPS URL of your git repository from step 1
+- Credentials: Use the drop-down menu to choose the credentials you created in step 3
+- Branch Specifier: Remove the branch name from this so it will pickup changes from every branch.
+- Under **Build Triggers**, check the box next to **Poll SCM**, then enter `* * * * *` in the corresponding text box. This is a cron expression that will check the repository for a change every minute.
```
Comment (chris-s-friedman) [inline exercises/03-Polling-SCM/README.md]: up to you, just thinking about the unintiated who may not know what cron is - you may want to drop a click link to wikipedia or have a quick definition of what cron is e.g. `cron is a tool used to schedule the execution of tasks based on time`

File: exercises/03-Polling-SCM/README.md [hunk 12]
```diff
@@ -1,8 +1,96 @@
...
+
```
Comment (chris-s-friedman) [inline exercises/03-Polling-SCM/README.md:62]: may want to drop a note here to hit "save"

File: exercises/03-Polling-SCM/README.md [hunk 13]
```diff
@@ -1,8 +1,96 @@
...
+**Step 5: Poll Github from a Pipeline Job**
+-------------------------------------------
+
+Pipeline jobs are a little stricter in terms of what they'll allow configuration-wise. Instead of running the pipeline script from Jenkins as you've been doing, you're going to check it into SCM now.
+
+Go to the pipeline job in your Jenkins console and open the **Configure** page. From here, check the box next to **Poll SCM**, then enter `* * * * *` in the corresponding text box.
+
+Further down in the **Pipeline** section, change `Pipeline Script` to `Pipeline Script from SCM`, then fill out the parameters below.
```
Comment (chris-s-friedman) [inline exercises/03-Polling-SCM/README.md:72]: after selecting `Pipeline Script from SCM`, you also need to indicate that the scm is git

File: exercises/03-Polling-SCM/README.md [hunk 14]
```diff
@@ -1,8 +1,96 @@
...
+
+- Repository URL: The HTTPS URL of your git repository from step 1
+- Credentials: Use the drop-down menu to choose the credentials you created in step 3
+- Branch Specifier: `dev`
+
+> Pipeline jobs can't run on each branch, only one at a time. If you need to run on more than one branch at a time, you need a Multi-Branch pipeline job, which we'll be covering later in the course.
+
```
Comment (chris-s-friedman) [inline exercises/03-Polling-SCM/README.md:81]: again, i  assume i should hit `save` here?

File: exercises/03-Polling-SCM/README.md [hunk 15]
```diff
@@ -1,8 +1,96 @@
...
+**Step 6: Checkin the Jenkinsfile**
+-------------------------------------------
+
+Take the file named `Jenkinsfile` from this folder and copy it to the github repository you are polling in the Freestyle Job. 
+
+Then from your git repository, run these commands:
```
Comment (chris-s-friedman) [inline exercises/03-Polling-SCM/README.md:87]: for folks less used to git, not used to the command line, this may prove to be a sticky part. 😬

File: exercises/03-Polling-SCM/README.md [hunk 16]
```diff
@@ -1,8 +1,96 @@
...
+- Repository URL: The HTTPS URL of your git repository from step 1
+- Credentials: Use the drop-down menu to choose the credentials you created in step 3
+- Branch Specifier: `dev`
+
+> Pipeline jobs can't run on each branch, only one at a time. If you need to run on more than one branch at a time, you need a Multi-Branch pipeline job, which we'll be covering later in the course.
+
+**Step 6: Checkin the Jenkinsfile**
+-------------------------------------------
+
+Take the file named `Jenkinsfile` from this folder and copy it to the github repository you are polling in the Freestyle Job.
```
Comment (chris-s-friedman) [inline exercises/03-Polling-SCM/README.md]: you may want to drop an indication that this is the repo we created in step 1 and not the clone of this repo

File: exercises/03-Polling-SCM/README.md [hunk 17]
```diff
@@ -1,8 +1,96 @@
...
+- `git push --set-upstream origin dev`
+
+Those commands do the following:
+
+- Create a new branch called `dev`
+- Stage the new Jenkinsfile
+- Commit the Jenkinsfile to the new `dev` branch
+- Push the dev branch with the new commit to the remote repository
+
+Once you push the branch, buth the Hello World Freestyle job and the Pipeline job should both trigger!
```
Comment (chris-s-friedman) [inline exercises/03-Polling-SCM/README.md:100]: I only got the freestyle job to trigger 😦

**Review by chris-s-friedman — COMMENTED** [review]
reviewed the scm stuff!

File: exercises/00-Environment-Setup/README.md (hunk 2 above)
Comment (devbyaccident) [reply exercises/00-Environment-Setup/README.md:73]: I tried a few different times and couldn't replicate this... I am not sure what happened, but I think this might have been a 1-off.

File: exercises/00-Environment-Setup/README.md (hunk 3 above)
Comment (devbyaccident) [reply exercises/00-Environment-Setup/README.md:82]: Fixed in https://github.com/devbyaccident/Jenkins-level-1/pull/1/commits/c07266478e5092fb1fb3b1c8382ecc8f3dec7baa

File: exercises/01-Installing-Plugins/README.md
Comment (devbyaccident) [reply exercises/01-Installing-Plugins/README.md:1]: I think this is related to the 1-off from earlier. Will resolve this.

File: exercises/01-Installing-Plugins/README.md (hunk 4 above)
Comment (devbyaccident) [reply exercises/01-Installing-Plugins/README.md]: Heh, there's another, separate plugin for github branch source... But this is a good point, I should add in something about unique identifiers.

File: exercises/04-Build-Agents/README.md [hunk 18]
```diff
@@ -1,8 +1,76 @@
...
+You can setup a new node agent by navigating to **Manage Jenkins** > **Nodes** and clicking on **New Node**.
+
+Give it the name `docker-agent` and select the radio button to make it a **Permenant Agent**, then click **Create**
+
+On the next window, fill out the following parameters and leave the rest as they are:
+
+- **Remote root directory**: `/var/jenkins`
+- **Labels**: `agent`
+
+Click **Save, then from the nodes window, click on the new node. That should show you the commands to connect it to the controller via JNLP once you have another node running on the same network. Save the commands to connect a Unix agent and proceed to the next step.
```
Comment (chris-s-friedman) [inline exercises/04-Build-Agents/README.md]: fix the bold around `save`
```suggestion
Click **Save**, then from the nodes window, click on the new node. That should show you the commands to connect it to the controller via JNLP once you have another node running on the same network. Save the commands to connect a Unix agent and proceed to the next step.
```

File: exercises/04-Build-Agents/README.md [hunk 19]
```diff
@@ -1,8 +1,76 @@
...
+
+> Note: Some of the commands in this exercise will change `localhost` to the name of a container in the same network. You ideally won't have to do that in a real Jenkins environment that has been configured correctly, but it's unfortunatley unavoidable in a lab such as this. Please let the instructor know if you need any help with this!
+
+Start by opening a shell on the agent using the command below:
+
+`docker exec -it agent /bin/bash`
+
+You can get the docker container ID by running `hostname` from this shell window. Save that for later, you'll need it in a later step.
+
+Pull up the commands from the end of step 1, and replace all instances of `localhost` with `jenkins`. This is because that is the container name of the jenkins controller defined in the docker compose file, and the agent will be able to find it on the docker network.
```
Comment (chris-s-friedman) [inline exercises/04-Build-Agents/README.md]: it's unclear what commands you're talking about here, so it's hard to follow the rest of this guide

File: exercises/03-Polling-SCM/README.md (hunk 7 above)
Comment (devbyaccident) [reply exercises/03-Polling-SCM/README.md]: > What caught my eye was using the word clone - are you using that for an audience that would be unfamiliar with the term fork?

Good catch, thanks,

> The other thing I'm thinking about - and take this with the grain of salt that I'm commenting before reading the rest of this page - shouldn't this really say git repo on github instead of just git repo?

That might be adding too much additional info. Trying to keep this simple to start.

File: exercises/03-Polling-SCM/README.md (hunk 15 above)
Comment (devbyaccident) [reply exercises/03-Polling-SCM/README.md:87]: Yea, that's part of the course though... Might be worth re-working in the future.
